speculator/train_speculator.py

```diff
@@ -0,0 +1,321 @@
+
+
+def test_model(rank, model, arch, cfg, prompt_type="chat"):
+    if rank == 0:
+        print("testing model output")
+    tokenizer = tokenizers.get_tokenizer(cfg.model_path)
+    if prompt_type == "chat":
+        template = "Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\n{}\n\n### Response:"
+        prompt = template.format(
+            "Provide a list of instructions for preparing chicken soup."
+        )
+    else:
+        template = "[INST] Write code to solve the following coding problem that obeys the constraints and passes the example test cases. Please wrap your code answer using ```:\n{}\n[/INST]"
+        prompt = template.format("Write a bubble sort function in python.")
+
+    tokens = tokenizer.tokenize(prompt)
+    ids = tokenizer.convert_tokens_to_ids(tokens)
+    if "llama" in arch:
+        ids = [tokenizer.bos_token_id] + ids
+    ids = torch.tensor(ids, dtype=torch.long, device="cuda")
+    result = generation.generate(
+        model,
+        ids,
+        max_new_tokens=100,
+        use_cache=True,
+        do_sample=False,
+        max_seq_len=8192,
+    )
+    result = generation.truncate_after_eos(result, tokenizer.eos_token_id)
+    if rank == 0:
+        print(f"{rank}: quick test of base model")
+        print(
+            tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(result))
+        )
+
+
+def main(**kwargs):
+    # get configs
+    cfg = config.train_config()
+    update_config(cfg, **kwargs)
+    cfg.seq_length = cfg.seq_length + cfg.n_speculator_heads + 1
+
+    # ensure reproducibility
+    torch.cuda.manual_seed(cfg.seed)
+    torch.manual_seed(cfg.seed)
+
+    # torchrun specific
+    local_rank = int(os.environ["LOCAL_RANK"])
+    rank = int(os.environ["RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+
+    if rank == 0:
+        print(f"{time.time()} running with these configs {cfg}")
+
+    # some setups
+    torch.cuda.set_device(local_rank)
+
+    if cfg.sharding_strategy != "tp":
+        setup()
+        torch._C._distributed_c10d._register_process_group("default", dist.group.WORLD)
+        base_model_mesh = None
+        speculator_mesh = None
+    else:
+        tp_size = 8
+        base_model_mesh = dist.device_mesh.init_device_mesh(
+            "cuda", (world_size // tp_size, tp_size), mesh_dim_names=("dp", "tp")
+        )
+        speculator_mesh = dist.device_mesh.init_device_mesh("cuda", (world_size,))
+        torch._C._distributed_c10d._register_process_group(
+            "default", base_model_mesh["tp"].get_group()
+        )
+
+    torch.cuda.empty_cache()
+    setup_environ_flags()
+    torch.set_default_dtype(torch.bfloat16)
+
+    # get policy
+    (
+        mixed_precision_policy,
+        wrapping_policy,
+        sharding_strategy_policy,
+        apply_selective_ac,
+        param_init_fn,
+    ) = get_policies(cfg, rank, LLaMABlock)
+
+    model = get_model(
```

Comment:
Currently we are assuming source="hf", might make sense to try the new `get_model` from https://github.com/foundation-model-stack/foundation-model-stack/pull/326

Reply:
Will include that in next iteration after testing